physical/raft/raft.go

```diff
@@ -2006,3 +2272,10 @@ func etcdboltOptions(path string) *etcdbolt.Options {
 
 	return o
 }
+
+func isRaftLogVerifyCheckpoint(l *raft.Log) bool {
+	return len(l.Data) == 1 &&
+		l.Data[0] == byte(verifierCheckpointOp) &&
+		len(l.Extensions) >= 8 &&
+		bytes.Equal(logVerifierMagicBytes[:], l.Extensions[0:8])
```

Comment:
As discussed offline, I was wrong in my guidance here and this will cause problems!

I was write that it could cause collisions with non-checkpoints before so needed fixing (Consul's implementation still can as far as I can see, although is likely saved in practice by most message encodings not being able to generate a collision. Still I'll file an issue).

The problem here is that I was assuming that we wrote the magic extensions prefix at the time we sent the checkpoint to disambuguate. But even if we did, that would break because the verify currently expects the leader to write a checkpoint with empty extensions: https://github.com/hashicorp/raft-wal/blob/94a96b0533002d094b74e6d1bfbf56ec0e1b3c80/verifier/store.go#L135

So when this `IsCheckpointFn` is run on a leader now it will fail and the checkpoint will be passed to lower layers where it will cause issues as it's not something they understand.

Luckily there is a simple solution that still fixes the original issue I spotted:

```suggestion
        if !bytes.Equal(l.Data, []byte{verifierCheckpointOp}) {
            return false
        }

        // Single byte log with that byte value can only be a checkpoint or 
        // the last byte of a chunked message. If it's chunked it will have
        // chunking meta data.
        if len(l.Extensions) == 0 {
            // No meta data, must be a checkpoint on the leader with no 
            // verifier meta data yet.
            return true
        }
        if bytes.HasPrefix(l.Extensions, logVerifierMagicBytes[:]) {
           // Has verifier meta data so must be a replicated checkpoint on a follower
           return true
        }
        // Must be the last chunk of a chunked object that has chunking meta
        return false
```

I'm pretty sure this works in all cases and should fix the tests again. If you want to test this last part is necessary you could revert to how it was in Consul (just the `l.Data` checks) and then add a test that attempts to write something 1 byte larger than the chunk size in Vault and whose final byte is a `verifierCheckpointOp`.

Reply:
Here's what I ended up with:

```go
func isRaftLogVerifyCheckpoint(l *raft.Log) bool {
	chkpnt := make([]byte, 1)
	chkpnt[0] = byte(verifierCheckpointOp)

	if !bytes.Equal(l.Data, chkpnt) {
		return false
	}

	// Single byte log with that byte value can only be a checkpoint or
	// the last byte of a chunked message. If it's chunked it will have
	// chunking metadata.
	if len(l.Extensions) == 0 {
		// No metadata, must be a checkpoint on the leader with no
		// verifier metadata yet.
		return true
	}

	if bytes.HasPrefix(l.Extensions, logVerifierMagicBytes[:]) {
		// Has verifier metadata so must be a replicated checkpoint on a follower
		return true
	}

	// Must be the last chunk of a chunked object that has chunking meta
	return false
}
```

It's identical to yours except that `verifierCheckpointOp` is a uint32 (because all of the other pre-existing raft ops are uint32s) so you can't drop it directly into a byte slice. In fact, when we write the checkpoint op into the raft data, we cast it to a byte, which truncates the 3 zero bytes at the end. Here is that code:

```go
func (b *RaftBackend) applyVerifierCheckpoint() error {
	data := make([]byte, 1)
	data[0] = byte(verifierCheckpointOp)

	b.permitPool.Acquire()
	b.l.RLock()

	var err error
	applyFuture := b.raft.Apply(data, 0)
	if e := applyFuture.Error(); e != nil {
		err = e
	}

	b.l.RUnlock()
	b.permitPool.Release()

	return err
}
```

So I'm doing the same thing at the top of the verifier function. With these changes, everything passes. Does that look ok to you?

Reply:
Yes! 👍 

You could do the byte case inline with my original suggestion too. Explicit make version is fine I can't say for sure but I suspect the compiler is smart enough to realise that it's not going to escape the function and so just use stack anyway so they probably are equivalent.

```go
if !bytes.Equal(l.Data, []byte{byte(verifierCheckpointOp)}) {
```

Reply:
Ah, copy that. I'll do that then - it's shorter.